### PR TITLE
compiler/next: Convert more internal modules to uAST

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
compiler/next: Convert more internal modules to uAST (#18721)

Convert more internal modules to `uAST` by working through a
combination of bugs and TODOs at parse/conversion time.

Changes follow:

Implement the conversion routine for the `foreach` loop, for the
statement level only. It looks like the grammar does not support
expression level `foreach` loops yet.

Implement the conversion routine for label statements.

Adjust expression-level try nodes to use a different builder that
generates a `PRIM_TRY` call instead of a block.

Add special handling to convert the call `foo.bytes()`. The previous
approach generated one too many calls.

Adjust one of the rules for parsing a bracket loop to properly
consume the loop iterand instead of the body twice. This fixed
a double free bug.

Properly reset the parser visibility level after parsing a `use` or
`import` statement.

Handle primary methods and leave a TODO to error if the receiver
for a primary method has a type expression.

Add special handling when converting extern types and config types.

Add another constructor for `astlocMarker` that takes a `Location`.
Care should be taken when pinning on `chpl::Location` over
`chpl::ID` as the location can grow stale when queries are rerun.
A better strategy could be used here in the future.

Set a global error handler for the compiler library context in
the `uASTParseFile` function. For now, it simply dumps any parse
errors that are emitted with calls to `USR_FATAL_CONT` and
`USR_PRINT`. We may want to make this error handler more robust or
move it to another source location in the future.

Attach visibility to `Symbol` generated for `Variable`. We will
need to attach it to other generated symbols as future work.

Reviewed by @mppf, @aconsroe-hpe. Thanks!

TESTING:

  - [x] `ALL` on `linux64` when `COMM=none` (check that `chpl` works)
  - [x] All hellos with `--compiler-library-parser`
  - [x] All primers with `--compiler-library-parser`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

File: compiler/next/lib/parsing/ParserContextImpl.h [hunk 1]
```diff
@@ -32,6 +32,8 @@
 
 using chpl::types::Param;
 
+void gdbShouldBreakHere(void);
```
Comment (aconsroe-hpe) [inline compiler/next/lib/parsing/ParserContextImpl.h:35]: Is this related to the other changes in this file?

File: compiler/next/lib/uast/AggregateDecl.cpp [hunk 2]
```diff
@@ -30,8 +30,9 @@ bool AggregateDecl::validAggregateChildren(ASTListIteratorPair<Expression> it) {
     if (elt->isComment() || elt->isErroneousExpression()) {
       // OK
     } else if (elt->isDecl()) {
-      if (elt->isVariable() || elt->isFunction() ||
-          elt->isTupleDecl() || elt->isMultiDecl() ||
+      if (elt->isVariable() || elt->isFunction() || elt->isTupleDecl() ||
+          elt->isMultiDecl() ||
+          elt->isAggregateDecl() ||
```
Comment (aconsroe-hpe) [inline compiler/next/lib/uast/AggregateDecl.cpp:35]: I personally find the heavy nesting hard to follow (I know you're just updating existing work here and thats fine). It comes to mind that maybe a function like `validAggregateChild(Expression *)` could then be used with `std::all_of(it.begin(), it.end(), validAggregateChild)`.

File: compiler/next/lib/parsing/ParserContextImpl.h (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/next/lib/parsing/ParserContextImpl.h:35]: Oops, nope, I was using that to debug some stuff in the parser. Thanks!

File: compiler/parser/parser.cpp [hunk 3]
```diff
@@ -871,6 +930,17 @@ static ModuleSymbol* parseFile(const char* path,
   return retval;
 }
 
+// TODO: Add helpers to convert locations without passing IDs.
+static void uASTParseFileErrorHandler(const chpl::ErrorMessage& err) {
+  auto markError = astlocMarker(err.location());
```
Comment (aconsroe-hpe) [inline compiler/parser/parser.cpp:935]: Are you using `auto` over `astlocMarker markError(err.location())` for a specific purpose?

File: compiler/passes/convert-uast.cpp [hunk 4]
```diff
@@ -1163,6 +1215,15 @@ struct Converter {
       } else if (name == borrowedStr) {
         ret = new CallExpr(PRIM_TO_BORROWED_CLASS_CHECKED);
       }
+
+    // Unwrap 'foo.bytes' specially to avoid wrapping it in a third call.
```
Comment (aconsroe-hpe) [inline compiler/passes/convert-uast.cpp:1219]: Is this specific to `bytes` somehow? Are all other dot expressions handled so they don't create a third call?

File: compiler/passes/convert-uast.cpp [hunk 5]
```diff
@@ -1663,10 +1748,12 @@ struct Converter {
     }
 
     Expr* retType = nullptr;
-    if (node->returnType() && node->returnType()->isBracketLoop()) {
-      retType = convertArrayType(node->returnType()->toBracketLoop());
-    } else {
-      retType = convertExprOrNull(node->returnType());
+    if (node->returnType()) {
```
Comment (aconsroe-hpe) [inline compiler/passes/convert-uast.cpp:1751]: could you use `if (Expression *returnType = node->returnType()) { ...`?

File: compiler/parser/parser.cpp (hunk 3 above)
Comment (dlongnecke-cray) [reply compiler/parser/parser.cpp:935]: No, I just prefer to write it that way, but that might be C++'tastic so I'll defer to your judgement. I was worried at one point that it might introduce a copy but I tested with MSVC (perhaps not the best metric lol) and it didn't seem like that was the case?

Comment (mppf) [reply compiler/parser/parser.cpp:935]: I also tend to like this auto version. I've seen some C++ talks about how the auto version reduces a syntactic confusion / adds consistency because you can use the same pattern with a zero-args constructor e.g. `auto mark = astLocMarker();` where `astlocMarker mark();` is an error.

Comment (mppf) [reply compiler/parser/parser.cpp:935]: (FWIW it is also more similar to Chapel forms which people working on our compiler are probably familiar with).

Comment (mppf) [reply compiler/parser/parser.cpp:935]: Apart from that though I'm not seeing why `astlocMarker` is used here. I would expect instead that the Location is passed as the first argument to `USR_FATAL_CONT` / `USR_PRINT` etc. The `astlocMarker` is updating a global sortof last resort location that is used when constructing AST nodes most of all.

File: compiler/parser/parser.cpp [hunk 6]
```diff
@@ -882,6 +952,8 @@ static ModuleSymbol* uASTParseFile(const char* fileName,
     INT_FATAL("compiler library context not initialized");
   }
 
+  gContext->setErrorHandler(&uASTParseFileErrorHandler);
```
Comment (mppf) [inline compiler/parser/parser.cpp:955]: IMO we should do this when we set up gContext so we have this set up across the whole compiler (which is only relevant when we go beyond what we have here).

File: compiler/passes/convert-uast.cpp [hunk 7]
```diff
@@ -650,6 +665,23 @@ struct Converter {
...
             : nullptr;
 
+      Expr* cond = nullptr;
+
+      // TODO: Can 'ifVars' happen in expression-level conditionals?
+      if (auto ifVar = node->condition()->toVariable()) {
+        assert(ifVar->kind() == uast::Variable::CONST ||
+               ifVar->kind() == uast::Variable::VAR);
+        assert(ifVar->initExpression());
+        auto identDef = ifVar->name().c_str();
```
Comment (mppf) [inline compiler/passes/convert-uast.cpp:675]: This is the name of it, not a DefExpr, so suggest renaming to e.g. varName.

File: compiler/passes/convert-uast.cpp (hunk 4 above)
Comment (mppf) [reply compiler/passes/convert-uast.cpp:1219]: This code is just handling the situation that we are calling a keyword, which isn't represented in the normal way, in the production AST anyway. Other  regular dot expressions are handled elsewhere in this file (David could tell you where).

**Review by mppf — APPROVED** [review]

File: compiler/passes/convert-uast.cpp (hunk 4 above)
Comment (aconsroe-hpe) [reply compiler/passes/convert-uast.cpp:1219]: TIL `bytes` gets its own token

Comment (mppf) [reply compiler/passes/convert-uast.cpp:1219]: Yeah, we have many of the builtin types have their own token so we treat them as "keywords". Mainly we don't want people to do things like `var bytes = 23` but it is of course debateable whether handling that in the parser was the right move.

File: compiler/parser/parser.cpp (hunk 3 above)
Comment (dlongnecke-cray) [reply compiler/parser/parser.cpp:935]: I just followed the pattern used in the dispatcher in `convert-uast.cpp`, but I can look into passing the location to `USR_*` stuff as a followup?

File: compiler/parser/parser.cpp (hunk 6 above)
Comment (dlongnecke-cray) [reply compiler/parser/parser.cpp:955]: I'll add a TODO

File: compiler/passes/convert-uast.cpp (hunk 7 above)
Comment (dlongnecke-cray) [reply compiler/passes/convert-uast.cpp:675]: Should I also change the arg names of the old parser helpers? That's what I've been following, but perhaps not the most sound reasoning.

Comment (mppf) [reply compiler/passes/convert-uast.cpp:675]: I think it'd be reasonable to do so but it's not a requirement

**Comment by dlongnecke-cray** [comment]
Some errors to attend to:

```
  2 [Error matching compiler output for release/examples/primers/chplvis/chplvis2]
  3 [Error matching compiler output for release/examples/primers/chplvis/chplvis3]
  4 [Error matching compiler output for release/examples/primers/chplvis/chplvis4]
  5 [Error matching compiler output for release/examples/primers/LinearAlgebralib]
  6 [Error matching compiler output for release/examples/primers/arrays]
  7 [Error matching compiler output for release/examples/primers/associative]
  8 [Error matching compiler output for release/examples/primers/distributions]
  9 [Error matching compiler output for release/examples/primers/domains]
 10 [Error matching compiler output for release/examples/primers/fileIO]
 11 [Error matching compiler output for release/examples/primers/forallLoops]
 12 [Error matching compiler output for release/examples/primers/learnChapelInYMinutes]
 13 [Error matching compiler output for release/examples/primers/parIters]
 14 [Error matching compiler output for release/examples/primers/procedures]
 15 [Error matching compiler output for release/examples/primers/randomNumbers]
 16 [Error matching compiler output for release/examples/primers/ranges]
 17 [Error matching compiler output for release/examples/primers/reductions]
 18 [Error matching compiler output for release/examples/primers/replicated]
 19 [Error matching compiler output for release/examples/primers/slices]
 20 [Error matching compiler output for release/examples/primers/sparse]
 21 [Error matching compiler output for release/examples/primers/specialMethods]
 22 [Error matching compiler output for release/examples/primers/syncsingle]
 23 [Error matching compiler output for release/examples/primers/tuples]
 24 [Error matching compiler output for release/examples/primers/variables]
 25 [Summary: #Successes = 16 | #Failures = 23 | #Futures = 0]
 26 [END]
```

**Comment by dlongnecke-cray** [comment]
I've commented out internal modules until things work again.
